File: packages/typescript/examples/validators-app.ts
#!/usr/bin/env npx ts-node
/**
 * Validators App - TypeScript
 *
 * Demuestra los validadores de RFC, CURP, CLABE y NSS.
 *
 * Uso:
 *   npx ts-node examples/validators-app.ts
 */

import { validateRfc, getRfcInfo, generateRfcFisica, generateRfcMoral } from '../src/validators/rfc';
import { validateCurp, getCurpInfo, generateCurp } from '../src/validators/curp';
import { validateClabe, getClabeInfo, generateClabe } from '../src/validators/clabe';
import { validateNss, getNssInfo, generateNss } from '../src/validators/nss';

function printHeader(): void {
  console.log();
  console.log('╔══════════════════════════════════════════════════════╗');
  console.log('║     🪪 Validadores Mexicanos - catalogmx             ║');
  console.log('╚══════════════════════════════════════════════════════╝');
  console.log();
}

function demoRfc(): void {
  console.log('  📋 RFC (Registro Federal de Contribuyentes)');
  console.log('  ───────────────────────────────────────────');
  console.log();

  // Generar RFC persona física
  const rfcFisica = generateRfcFisica({
    nombre: 'Juan',
    apellidoPaterno: 'García',
    apellidoMaterno: 'López',
    fechaNacimiento: new Date(1990, 4, 15), // 15 mayo 1990
  });

  console.log('  Persona Física:');
  console.log(`    Nombre: Juan García López`);
  console.log(`    Fecha:  1990-05-15`);
  console.log(`    RFC:    ${rfcFisica}`);
  console.log(`    Válido: ${validateRfc(rfcFisica) ? '✅' : '❌'}`);

  const infoFisica = getRfcInfo(rfcFisica);
  if (infoFisica) {
    console.log(`    Tipo:   ${infoFisica.type}`);
  }
  console.log();

  // Generar RFC persona moral
  const rfcMoral = generateRfcMoral({
    razonSocial: 'Tecnologías Avanzadas',
    fechaConstitucion: new Date(2010, 5, 1), // 1 junio 2010
  });

  console.log('  Persona Moral:');
  console.log(`    Razón Social: Tecnologías Avanzadas`);
  console.log(`    Constitución: 2010-06-01`);
  console.log(`    RFC:          ${rfcMoral}`);
  console.log(`    Válido:       ${validateRfc(rfcMoral) ? '✅' : '❌'}`);
  console.log();
}

function demoCurp(): void {
  console.log('  🆔 CURP (Clave Única de Registro de Población)');
  console.log('  ──────────────────────────────────────────────');
  console.log();

  // Generar CURP
  const curp = generateCurp({
    nombre: 'María',
    apellidoPaterno: 'Hernández',
    apellidoMaterno: 'Sánchez',
    fechaNacimiento: new Date(1985, 2, 20), // 20 marzo 1985
    sexo: 'M',
    estado: 'Jalisco',
  });

  console.log('  Datos:');
  console.log(`    Nombre: María Hernández Sánchez`);
  console.log(`    Fecha:  1985-03-20`);
  console.log(`    Sexo:   Mujer`);
  console.log(`    Estado: Jalisco`);
  console.log();
  console.log(`    CURP:   ${curp}`);
  console.log(`    Válido: ${validateCurp(curp) ? '✅' : '❌'}`);

  const info = getCurpInfo(curp);
  if (info) {
    console.log(`    Género: ${info.gender === 'M' ? 'Mujer' : 'Hombre'}`);
    console.log(`    Estado: ${info.stateCode}`);
  }
  console.log();
}

function demoClabe(): void {
  console.log('  🏦 CLABE (Clave Bancaria Estandarizada)');
  console.log('  ───────────────────────────────────────');
  console.log();

  // Generar CLABE
  const clabe = generateClabe('012', '180', '12345678901');

  console.log('  Datos:');
  console.log(`    Banco:   012 (BBVA México)`);
  console.log(`    Plaza:   180`);
  console.log(`    Cuenta:  12345678901`);
  console.log();
  console.log(`    CLABE:   ${clabe}`);
  console.log(`    Válida:  ${validateClabe(clabe) ? '✅' : '❌'}`);

  const info = getClabeInfo(clabe);
  if (info) {
    console.log(`    Banco:   ${info.bankCode}`);
    console.log(`    Check:   ${info.checkDigit}`);
  }
  console.log();
}

function demoNss(): void {
  console.log('  🏥 NSS (Número de Seguridad Social)');
  console.log('  ────────────────────────────────────');
  console.log();

  // Generar NSS
  const nss = generateNss(45, 90, 85, 1234);

  console.log('  Datos:');
  console.log(`    Subdelegación:   45`);
  console.log(`    Año registro:    90`);
  console.log(`    Año nacimiento:  85`);
  console.log(`    Secuencial:      1234`);
  console.log();
  console.log(`    NSS:     ${nss}`);
  console.log(`    Válido:  ${validateNss(nss) ? '✅' : '❌'}`);

  const info = getNssInfo(nss);
  if (info) {
    console.log(`    Subdel:  ${info.subdelegacion}`);
  }
  console.log();
}

function validarCodigo(codigo: string): void {
  console.log(`  Validando: ${codigo}`);
  console.log('  ─────────────────────────────────');

  const len = codigo.length;

  if (len === 18) {
    // Puede ser CURP o CLABE
    if (/^\d+$/.test(codigo)) {
      // CLABE (solo dígitos)
      const esValido = validateClabe(codigo);
      console.log(`  Tipo:   CLABE`);
      console.log(`  Válido: ${esValido ? '✅' : '❌'}`);
      if (esValido) {
        const info = getClabeInfo(codigo);
        console.log(`  Banco:  ${info?.bankCode}`);
      }
    } else {
      // CURP
      const esValido = validateCurp(codigo);
      console.log(`  Tipo:   CURP`);
      console.log(`  Válido: ${esValido ? '✅' : '❌'}`);
    }
  } else if (len === 13) {
    // RFC persona física
    const esValido = validateRfc(codigo);
    console.log(`  Tipo:   RFC (persona física)`);
    console.log(`  Válido: ${esValido ? '✅' : '❌'}`);
  } else if (len === 12) {
    // RFC persona moral
    const esValido = validateRfc(codigo);
    console.log(`  Tipo:   RFC (persona moral)`);
    console.log(`  Válido: ${esValido ? '✅' : '❌'}`);
  } else if (len === 11) {
    // NSS
    const esValido = validateNss(codigo);
    console.log(`  Tipo:   NSS`);
    console.log(`  Válido: ${esValido ? '✅' : '❌'}`);
  } else {
    console.log(`  ❓ Código no reconocido (longitud: ${len})`);
  }
  console.log();
}

// Main
async function main(): Promise<void> {
  printHeader();

  const args = process.argv.slice(2);
  const comando = args[0]?.toLowerCase() || 'demo';

  switch (comando) {
    case 'demo':
      demoRfc();
      demoCurp();
      demoClabe();
      demoNss();
      break;

    case 'rfc':
      demoRfc();
      break;

    case 'curp':
      demoCurp();
      break;

    case 'clabe':
      demoClabe();
      break;

    case 'nss':
      demoNss();
      break;

    case 'validar':
      if (args[1]) {
        validarCodigo(args[1].toUpperCase());
      } else {
        console.log('  Uso: npx ts-node validators-app.ts validar <código>');
      }
      break;

    default:
      console.log('  Uso:');
      console.log('    npx ts-node validators-app.ts demo');
      console.log('    npx ts-node validators-app.ts rfc');
      console.log('    npx ts-node validators-app.ts curp');
      console.log('    npx ts-node validators-app.ts clabe');
      console.log('    npx ts-node validators-app.ts nss');
      console.log('    npx ts-node validators-app.ts validar GALO850320HJCRPN07');
  }
}

main().catch(console.error);

File: packages/dart/example/validators_app.dart
#!/usr/bin/env dart
/// Validators App - Dart
///
/// Demuestra los validadores de RFC, CURP, CLABE y NSS.
///
/// Uso:
///   dart run example/validators_app.dart
library;

import 'package:catalogmx/catalogmx.dart';

void printHeader() {
  print('');
  print('╔══════════════════════════════════════════════════════╗');
  print('║     🪪 Validadores Mexicanos - catalogmx             ║');
  print('╚══════════════════════════════════════════════════════╝');
  print('');
}

void demoRfc() {
  print('  📋 RFC (Registro Federal de Contribuyentes)');
  print('  ───────────────────────────────────────────');
  print('');

  // Generar RFC persona física
  final rfcFisica = generateRFC(
    nombre: 'Juan',
    apellidoPaterno: 'García',
    apellidoMaterno: 'López',
    fechaNacimiento: DateTime(1990, 5, 15),
  );

  print('  Persona Física:');
  print('    Nombre: Juan García López');
  print('    Fecha:  1990-05-15');
  print('    RFC:    $rfcFisica');
  print('    Válido: ${validateRFC(rfcFisica) ? '✅' : '❌'}');
  print('');

  // Generar RFC persona moral
  final rfcMoral = generateRFCMoral(
    razonSocial: 'Tecnologías Avanzadas',
    fechaConstitucion: DateTime(2010, 6, 1),
  );

  print('  Persona Moral:');
  print('    Razón Social: Tecnologías Avanzadas');
  print('    Constitución: 2010-06-01');
  print('    RFC:          $rfcMoral');
  print('    Válido:       ${validateRFC(rfcMoral) ? '✅' : '❌'}');
  print('');
}

void demoCurp() {
  print('  🆔 CURP (Clave Única de Registro de Población)');
  print('  ──────────────────────────────────────────────');
  print('');

  // Generar CURP
  final curp = generateCURP(
    nombre: 'María',
    apellidoPaterno: 'Hernández',
    apellidoMaterno: 'Sánchez',
    fechaNacimiento: DateTime(1985, 3, 20),
    sexo: 'M',
    estado: 'Jalisco',
  );

  print('  Datos:');
  print('    Nombre: María Hernández Sánchez');
  print('    Fecha:  1985-03-20');
  print('    Sexo:   Mujer');
  print('    Estado: Jalisco');
  print('');
  print('    CURP:   $curp');
  print('    Válido: ${validateCURP(curp) ? '✅' : '❌'}');
  print('');
}

void demoClabe() {
  print('  🏦 CLABE (Clave Bancaria Estandarizada)');
  print('  ───────────────────────────────────────');
  print('');

  // Generar CLABE
  final clabe = generateCLABE(
    bankCode: '012',
    branchCode: '180',
    accountNumber: '12345678901',
  );

  print('  Datos:');
  print('    Banco:   012 (BBVA México)');
  print('    Plaza:   180');
  print('    Cuenta:  12345678901');
  print('');
  print('    CLABE:   $clabe');
  print('    Válida:  ${validateCLABE(clabe) ? '✅' : '❌'}');
  print('');
}

void demoNss() {
  print('  🏥 NSS (Número de Seguridad Social)');
  print('  ────────────────────────────────────');
  print('');

  // Generar NSS
  final nss = generateNSS(
    subdelegacion: 45,
    anioRegistro: 90,
    anioNacimiento: 85,
    secuencial: 1234,
  );

  print('  Datos:');
  print('    Subdelegación:   45');
  print('    Año registro:    90');
  print('    Año nacimiento:  85');
  print('    Secuencial:      1234');
  print('');
  print('    NSS:     $nss');
  print('    Válido:  ${validateNSS(nss) ? '✅' : '❌'}');
  print('');
}

void validarCodigo(String codigo) {
  print('  Validando: $codigo');
  print('  ─────────────────────────────────');

  final len = codigo.length;

  if (len == 18) {
    // Puede ser CURP o CLABE
    if (RegExp(r'^\d+$').hasMatch(codigo)) {
      // CLABE (solo dígitos)
      final esValido = validateCLABE(codigo);
      print('  Tipo:   CLABE');
      print('  Válido: ${esValido ? '✅' : '❌'}');
    } else {
      // CURP
      final esValido = validateCURP(codigo);
      print('  Tipo:   CURP');
      print('  Válido: ${esValido ? '✅' : '❌'}');
    }
  } else if (len == 13) {
    // RFC persona física
    final esValido = validateRFC(codigo);
    print('  Tipo:   RFC (persona física)');
    print('  Válido: ${esValido ? '✅' : '❌'}');
  } else if (len == 12) {
    // RFC persona moral
    final esValido = validateRFC(codigo);
    print('  Tipo:   RFC (persona moral)');
    print('  Válido: ${esValido ? '✅' : '❌'}');
  } else if (len == 11) {
    // NSS
    final esValido = validateNSS(codigo);
    print('  Tipo:   NSS');
    print('  Válido: ${esValido ? '✅' : '❌'}');
  } else {
    print('  ❓ Código no reconocido (longitud: $len)');
  }
  print('');
}

void main(List<String> args) {
  printHeader();

  final comando = args.isNotEmpty ? args[0].toLowerCase() : 'demo';

  switch (comando) {
    case 'demo':
      demoRfc();
      demoCurp();
      demoClabe();
      demoNss();
      break;

    case 'rfc':
      demoRfc();
      break;

    case 'curp':
      demoCurp();
      break;

    case 'clabe':
      demoClabe();
      break;

    case 'nss':
      demoNss();
      break;

    case 'validar':
      if (args.length > 1) {
        validarCodigo(args[1].toUpperCase());
      } else {
        print('  Uso: dart run example/validators_app.dart validar <código>');
      }
      break;

    default:
      print('  Uso:');
      print('    dart run example/validators_app.dart demo');
      print('    dart run example/validators_app.dart rfc');
      print('    dart run example/validators_app.dart curp');
      print('    dart run example/validators_app.dart clabe');
      print('    dart run example/validators_app.dart nss');
      print('    dart run example/validators_app.dart validar GALO850320HJCRPN07');
  }
}
